test/cql-pytest/test_cast.py

```diff
@@ -0,0 +1,165 @@
+# Compatible means that the bytes will stay unchanged after the conversion.
+# This means that it's legal to cast an int to blob (int is just a 4 byte blob),
+# but it's illegal to cast a bigint to int (change 4 bytes -> 8 bytes).
+# This simplifies things, to cast we can just reinterpret the value as the other type.
+
+# Another useful use of C-style casts is type hints.
+# Sometimes it's impossible to infer the exact type of an expression from the context.
+# In such cases the type can be specified by casting the expression to this type.
+# For example: `overloadedFunction((int)?)`
+# Without the cast it's impossible to guess what should be the bind marker's type.
+# The function is overloaded, so there are many possible argument types.
+# The type hint specifies that the bind marker has type int.
+
+# An interesting thing is that such casts don't have to be explicit.
+# CQL allows to put an int value in a place where a blob value is expected
+# and it will be automatically converted without any explicit casting.
+
+
+import pytest
+from pytest import fixture
+from cassandra.protocol import InvalidRequest
+from util import unique_name, unique_key_int
+import uuid
+
+@pytest.fixture(scope="module")
+def table1(cql, test_keyspace):
+    table = test_keyspace + "." + unique_name()
+    cql.execute(f"CREATE TABLE {table} (pk int PRIMARY KEY, blob_col blob, int_col int, bigint_col bigint)")
+    yield table
+    cql.execute("DROP TABLE " + table)
+
+@pytest.fixture(scope="module")
+def table2(cql, test_keyspace):
+    table = test_keyspace + "." + unique_name()
+    cql.execute(f"CREATE TABLE {table} (pk int PRIMARY KEY, d date)")
+    yield table
+    cql.execute("DROP TABLE " + table)
+
+# Test casting int values to blob values using C-style casts
+def test_cast_int_blob(cql, table1):
+    pk = unique_key_int()
+
+    # Implicitly casting an integer constant to blob fails, it's unkown what the exact type is - is it tinyint, bigint?
+    # It's important that it stays this way - in the future we might implement guessing the type of untyped constants
+    # by assiging the smallest type that can fit the constant, but I think that we shouldn't allow converting them
+    # to blob without explicitly specifying size of the integer.
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, 12)")
+
+    # Putting (blob) before the integer also fails, it's still unkown what the exact type is.
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, (blob)123)")
+
+    # Putting (int) before the integer specifies the exact type and it's possible to cast the value.
+    # This type of assignment isn't currently allowed, will work in the future.
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, (int)1234)")
```

Comment:
I don't think it should work.

Reply:
Such implicit casts are generally allowed, but I opened #13054 based on our previous discussion.
Disallowing implicit casts would be a breaking change. I guess we could allow it only in cases that worked previously, but it would be inconsistent and a nightmare to maintain the behavior for all the particular cases.

Reply:
Can you give some examples where implicit casts are done and not done currently?

The idea is, if we implicitly cast in many places and the change opens up a few more, it's not a big deal.

But if we implicitly cast in just a few places and the change opens it up to many more places, then we're worsening type checking significantly.

Reply:
btw, something like (blob)(int)1234 can work.